RFC: no functions in makeStyles() (#20651)
microsoft/fluentui: rfcs/convergence/make-styles-no-functions.md

};
```

This also removes the need of using the `useTheme()` hook for customers as tokens can be accessed directly, for example:

```tsx
import { tokens } from '@fluentui/react-theme';

function CustomComponent() {
  return <div style={{ color: tokens.borderRadiusNone /* is "var(--borderRadiusNone)" */ }} />;
}
```

That simplifies also integration with other CSS-in-JS from Fluent UI v8 or Fluent UI Northstar.

## Remove functions in `makeStyles()`

Once `tokens` are available there is no more need for functional style rules in `makeStyles()`:

```diff
import { makeStyles } from '@fluentui/react-make-styles';
+import { tokens } from '@fluentui/react-theme';

makeStyles({
-  root: theme => ({ color: theme.tokenB }),
+  root: { color: tokens.tokenB },
});
```

## Simplify types in `FluentProvider`

Currently `FluentProvider` only supports the `Theme` type from `@fluentui/react-theme`.

If we simplify this to `Record<string, string | number>`, we enable consumers to extend the default theme:

```tsx
import { FluentProvider } from '@fluentui/react-provider';
import { mergeThemes, teamsLightTheme, Theme } from '@fluentui/react-theme';

type CustomTokens = {
  tokenA: string;
};

> type CustomTheme = CustomTokens & Theme;

Reviewer: **thought (out of topic):** it would be nice if we could introduce more self explanatory API for consumers to create custom theme like following:

```ts
import type {CreateCustomThemeShape} from '@fui/react-theme'

type CustomTokens = {...}
type CustomTheme = CreateCustomThemeShape<CustomTokens >
```

or even more explicit, with less API surface needed

```ts
import { createCustomTheme } from '@fui/react-theme'

const customTokens = {...}
const customTheme = createCustomTheme(teamsLightTheme, customTokens)
// easily accessible shape inferred from source of truth - actual implementation
type CustomTheme = typeof customTheme
```

WDYT?

Reviewer 2: I like the second option and we could even make some changes to `mergeThemes` to allow for this type of functionality:

```ts
export function mergeThemes<TTheme extends Theme = Theme, TPartialTheme extends PartialTheme = PartialTheme>(a: TTheme | undefined, b: TPartialTheme | undefined): TTheme & TPartialTheme {
   ...
}
```

This would leverage our existing `mergeThemes` functionality so we don't have to create a separate `createCustomTheme` function.

WDYT of this approach? If you like it, I can go ahead and update the RFC with this info.

Reviewer 2: I'm spinning this up into a separate issue so we can continue the conversation without blocking this PR.

# Problem

> Currently `makeStyles()` implements two on how styles can be defined:

Reviewer 3: ```suggestion
Currently `makeStyles()` can define style rules in two different ways
```

```ts
makeStyles({
  rootA: { color: 'red' },
  rootB: theme => ({ color: theme.tokenB }),
});
```

> `theme` is typed strictly and offers tokens from `@fluentui/react-theme`. This tokens shape is not extensible for customers.

Reviewer 3: ```suggestion
`theme` is typed and coupled to tokens from `@fluentui/react-theme`. This tokens shape is not extensible for customers.
```

# Solution

## Export tokens separately

> Tokens are just CSS variables, we should be clear with customers on what they are using. This also removes need in `useTheme()` hook as tokens can be accessed directly.

Reviewer 3: ```suggestion
Tokens are just CSS variables, we communicate this fact to customers so that they understand clearly what they are using. This also removes need in `useTheme()` hook as tokens can be accessed directly.
```

```tsx
import { tokens } from '@fluentui/react-theme';

function CustomComponent() {
  return <div style={{ color: tokens.tokensA }} />;
}
```

`tokens` is just a plain object:

```ts
import type { Theme } from '@fluentui/react-theme';

const tokens: Theme = {
  borderRadiusNone: 'var(--borderRadiusNone)',
  borderRadiusSmall: 'var(--borderRadiusSmall)',
  /* ... */
};
```

## Remove functions

> Once `tokens` are available there is no sense to keep functions in `makeStyles()`:

Reviewer 3: ```suggestion
Once `tokens` are available there is no more need for functional style rules in `makeStyles()`:
```

```diff
import { makeStyles } from '@fluentui/react-make-styles';
+import { tokens } from '@fluentui/react-theme';

makeStyles({
-  root: theme => ({ color: theme.tokenB }),
+  root: { color: tokens.tokenB },
});
```

## Simplify types in `FluentProvider`

> Currently `FluentProvider` accepts `theme` only as `Theme` type from `@fluentui/react-theme`. If we will simplify it to `Record<string, string | number>` this will allow customers to extend our theme.

Reviewer 3: ```suggestion
Currently `FluentProvider` only supports the `Theme` type from `@fluentui/react-theme`. If we simplify this to `Record<string, string | number>` , we enable consumers to extend the default theme.
```

> Currently `FluentProvider` only supports the `Theme` type from `@fluentui/react-theme`. If we simplify this to `Record<string, string | number>` , we enable consumers to extend the default theme.

Reviewer 4: As `FluentProvider` currently accepts only the `Theme` type, it guarantees an interoperability between a host application and a component developed in isolation - any high value component used in any host application will have access to the expected design tokens.
With changing the type to a generic record, how can we enforce the contract?

Author: We can optional expose `TokensProvider` (`TokensToCSSProvider`, _whatever_):

- it will do only only `useThemeStyleTag()` part
- it will have `Record<string, ...>`
- keep `FluentProvider` with `Theme`

🔽🔽🔽

```tsx
function App(props) {
  return (
    <FluentProvider theme={teamsTheme}>
      <TokensProvider tokens={customTokens}>{props.children}</TokensProvider>
    </FluentProvider>
  );
}
```

Reviewer 4: Hmm, that might work.

Still think about the typings and the theme shape. There is one shape which you need to pass to the provider (`[cssVariableName; rawValue]` mapping) and another shape which you need to use in `makeStyles` (`[tokenName;--var(cssVariableName)]`).
Also there might be tokens which are not injected as CSS variables - those have no entries in the object you pass to the provider and have [token name; raw value] in the `makeStyles`.

Reviewer 2: I think there is another option here where `FluentProvider` accepts a `TTheme` that extends from `Theme`. Today, `FluentProviderProps` is typed as:

```ts
export interface FluentProviderProps
  extends Omit<ComponentProps<FluentProviderSlots>, 'dir'>,
    Partial<FluentProviderCommons> {
  theme?: PartialTheme;
}
```

We could make it be:

```ts
export interface FluentProviderProps<TTheme extends PartialTheme = PartialTheme>
  extends Omit<ComponentProps<FluentProviderSlots>, 'dir'>,
    Partial<FluentProviderCommons> {
  theme?: TTheme;
}
```

In this way, we are adding extensibility to the `FluentProvider` theme shape while still maintaining that interoperability you were talking about.

Author: @khmakoto This is also an option. We can export a primitive component that renders variables for `Record<string, string | number>` if there will be an ask for it.

Reviewer 2: I forgot to mention that I added snippets of this conversation to the RFC so that they are accounted for 😄

```tsx
import { FluentProvider } from '@fluentui/react-provider';
import { mergeThemes, teamsLightTheme, Theme } from '@fluentui/react-theme';

type CustomTokens = {
  tokenA: string;
};
type CustomTheme = CustomTokens & Theme;

const extendedTheme: CustomTheme = mergeThemes(teamsLightTheme, { tokenA: 'red' });

function App() {
  return <FluentProvider theme={extendedTheme} />;
}
```

> Usage of `FluentProvider` will inject all customer tokens properly and will make them available on React Portals.

Reviewer 3: ```suggestion
`FluentProvider` will inject all customer tokens properly. These tokens will also be available on React Portals.
```

## Using custom tokens in `makeStyles()`

> Once tokens are injected on `FluentProvider` they could be used in `makeStyles()`.

Reviewer 3: ```suggestion
Once tokens are injected through the `FluentProvider` they can be used in `makeStyles()`.
```
